File: android/app/src/main/kotlin/com/narraleaf/shell/NativeContentDecoder.kt
package com.narraleaf.shell

import java.io.FilterInputStream
import java.io.IOException
import java.io.InputStream

/**
 * [ContentDecoder] backed by the prebuilt decoder library.
 *
 * The library is a build input, not source: it arrives as a prebuilt `.so` per
 * ABI and exports the entry points declared below. This class is deliberately
 * the whole of the shell's knowledge on the subject — a header of
 * [nativeHeaderLen] bytes precedes the content, and content bytes are passed
 * through [nativeDecode] at their offset. What that does is not the shell's
 * business.
 *
 * The class and method names below are an ABI contract with the library:
 * renaming any of them breaks the JNI lookup, and it breaks it at runtime on a
 * device rather than at build time here.
 */
class NativeContentDecoder(private val key: String) : ContentDecoder {

    override fun decodedLength(sourceLength: Long): Long {
        // This is asked without the file in hand, so it leans on the packer
        // encoding the whole payload or none of it: a build with a key has a
        // header on every payload file. decode() checks each file individually
        // and fails loudly if that ever stops being true.
        val header = PREFIX_LEN.toLong()
        return if (sourceLength >= header) sourceLength - header else 0
    }

    override fun decode(source: InputStream, declaredLength: Long): InputStream {
        val head = ByteArray(PROBE_SIZE)
        var read = 0
        while (read < head.size) {
            val n = source.read(head, read, head.size - read)
            if (n < 0) break
            read += n
        }
        val headerLen = if (read < head.size) 0 else nativeHeaderLen(head)
        if (headerLen == 0) {
            source.close()
            // Serving the bytes as-is would hand the WebView ciphertext, which
            // fails far away from here (a blank game, a broken asset). This is
            // the packer and the shell disagreeing; say so.
            throw IOException("payload file is not encoded content")
        }
        if (headerLen != PREFIX_LEN) {
            // decodedLength() subtracts PREFIX_LEN without seeing the file. If
            // the library ever reports a different prefix than it asks to
            // probe, every Content-Length is silently wrong — refuse instead.
            source.close()
            throw IOException("decoder library reports an unexpected prefix ($headerLen != $PREFIX_LEN)")
        }
        val ctx = nativeOpen(head, key)
        if (ctx == 0L) {
            source.close()
            throw IOException("this build's key does not open its payload")
        }
        return DecodedStream(source, ctx)
    }

    /**
     * The decoded view of one file. Decodes lazily as it is read, and — the
     * load-bearing part — skips in O(1) by moving the offset instead of
     * decoding the bytes it passes over. Range requests seek by skipping, so
     * without this every seek into a long track would decode the whole track up
     * to that point.
     */
    private class DecodedStream(source: InputStream, private var ctx: Long) : FilterInputStream(source) {

        private var offset = 0L

        override fun read(): Int {
            val one = ByteArray(1)
            return if (read(one, 0, 1) < 0) -1 else one[0].toInt() and 0xff
        }

        override fun read(b: ByteArray, off: Int, len: Int): Int {
            val n = `in`.read(b, off, len)
            if (n <= 0) return n
            if (!nativeDecode(ctx, offset, b, off, n)) {
                throw IOException("decode failed")
            }
            offset += n.toLong()
            return n
        }

        override fun skip(n: Long): Long {
            val skipped = `in`.skip(n)
            if (skipped > 0) offset += skipped
            return skipped
        }

        override fun markSupported(): Boolean = false

        override fun close() {
            try {
                super.close()
            } finally {
                if (ctx != 0L) {
                    nativeFree(ctx)
                    ctx = 0L
                }
            }
        }
    }

    companion object {
        /**
         * The fixed number of leading bytes the library asks to see, which is
         * also the prefix it strips. decode() asserts the two agree per file
         * rather than trusting that they always will.
         */
        private val PROBE_SIZE: Int by lazy { nativeProbeSize() }
        private val PREFIX_LEN: Int get() = PROBE_SIZE

        /** Whether the library is present and loadable in this build. */
        val isAvailable: Boolean = try {
            System.loadLibrary("nlcrypto")
            true
        } catch (error: UnsatisfiedLinkError) {
            // A build with no payload key never links it, which is the normal
            // case for an unprotected game.
            false
        }

        @JvmStatic
        private external fun nativeProbeSize(): Int

        @JvmStatic
        private external fun nativeHeaderLen(head: ByteArray): Int

        @JvmStatic
        private external fun nativeOpen(head: ByteArray, obfKey: String): Long

        @JvmStatic
        private external fun nativeDecode(ctx: Long, contentOffset: Long, data: ByteArray, off: Int, len: Int): Boolean

        @JvmStatic
        private external fun nativeFree(ctx: Long)
    }
}

File: ios/Shell/NativeContentDecoder.swift
import Foundation
import NlCrypto

/// `ContentDecoder` backed by the prebuilt decoder library linked into the
/// shell binary.
///
/// The library is a build input, not source: it arrives as the prebuilt static
/// archive `NlCrypto.xcframework` and exports the `nlc_*` entry points its
/// header declares. This type is deliberately the whole of the shell's
/// knowledge on the subject — a file begins with a header of `nlc_header_len`
/// bytes, and content bytes are passed through `nlc_decode` at their offset.
/// What that does is not the shell's business.
///
/// The `nlc_*` names below are an ABI contract with the library; the header
/// ships inside the xcframework and the module map exposes it as `NlCrypto`.
final class NativeContentDecoder: ContentDecoder {

    private let key: String

    /// The fixed number of leading bytes the library asks to see, which is also
    /// the prefix it strips. `open` asserts the two agree per file rather than
    /// trusting that they always will.
    private static let probeSize: Int = nlc_probe_size()

    init(key: String) {
        self.key = key
    }

    func open(_ fileURL: URL, fileLength: UInt64) throws -> FileContentDecoder {
        let handle = try FileHandle(forReadingFrom: fileURL)
        defer { try? handle.close() }
        let head = (try handle.read(upToCount: Self.probeSize)) ?? Data()

        let headerLen: Int
        if head.count < Self.probeSize {
            headerLen = 0
        } else {
            headerLen = head.withUnsafeBytes { raw -> Int in
                guard let base = raw.baseAddress else { return 0 }
                return Int(nlc_header_len(base.assumingMemoryBound(to: UInt8.self), head.count))
            }
        }

        if headerLen == 0 {
            // A build with a key expects every payload file to carry a header.
            // Serving the bytes as-is would hand the web view content it cannot
            // use and fail far from here (a blank game, a broken asset); the
            // packer and the shell disagree, so say so.
            throw DecoderError.notEncoded
        }
        if headerLen != Self.probeSize {
            // `contentStart` leans on the prefix equalling the probe size. If
            // the library ever reports a different prefix than it probes, every
            // Content-Length would be silently wrong — refuse instead.
            throw DecoderError.unexpectedPrefix(headerLen)
        }

        let ctx: OpaquePointer? = head.withUnsafeBytes { raw in
            guard let base = raw.baseAddress else { return nil }
            return nlc_open(base.assumingMemoryBound(to: UInt8.self), head.count, key)
        }
        guard let ctx else {
            // A present but unusable key opens nothing. Refusing here keeps a
            // wrong build from serving content it cannot decode.
            throw DecoderError.keyRejected
        }
        return NativeFileContentDecoder(ctx: ctx, contentStart: UInt64(headerLen))
    }

    enum DecoderError: Error {
        case notEncoded
        case unexpectedPrefix(Int)
        case keyRejected
        case decodeFailed
        case closed
    }

    /// The decoded view of one file. Holds a native decoder for its lifetime and
    /// frees it on `close`. `nlc_decode` is safe to call concurrently on one
    /// decoder, so `decode` needs no lock of its own.
    private final class NativeFileContentDecoder: FileContentDecoder {

        let contentStart: UInt64
        private var ctx: OpaquePointer?
        private let lock = NSLock()

        init(ctx: OpaquePointer, contentStart: UInt64) {
            self.ctx = ctx
            self.contentStart = contentStart
        }

        func decode(_ data: Data, at contentOffset: UInt64) throws -> Data {
            guard let ctx else { throw DecoderError.closed }
            if data.isEmpty { return data }
            var out = data
            let ok = out.withUnsafeMutableBytes { raw -> Bool in
                guard let base = raw.baseAddress else { return false }
                return nlc_decode(ctx, contentOffset, base.assumingMemoryBound(to: UInt8.self), raw.count) == 1
            }
            guard ok else { throw DecoderError.decodeFailed }
            return out
        }

        func close() {
            lock.lock(); defer { lock.unlock() }
            if let ctx {
                nlc_free(ctx)
                self.ctx = nil
            }
        }
    }
}
